feat: enable STX by default with migration and notification

httpJunkie opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR enables Smart Transactions (STX) in MetaMask Mobile by default through migration number 68 for users who have either opted out with notification if the Smart Transactions toggle has been enabled for them. 

Docs: [SmartTransactionsMigrationBanner](https://www.notion.so/consensys/SmartTransactionsMigrationBanner-Feature-Documentation-180fc61a326e80a19da2ce55e6c8687f)

How it works (if user does not have STX enabled or prior STX Transaction history:

- Upon Migration STX is enabled in Settings:
 - A banner alert will displays on the following transaction confirmations
   - Send Confirmation flow
   - Swaps confirmation flow
   - Contract deployment & interactions (deploy, minting, etc.)

In the case a user migrates from a previous version of the Mobile app and the migration runs and sets STX toggle "ON" in `Settings > Advanced > Smart Transactions`, they will receive an Alert on transaction confirmation screens until dismissed, or by clicking on the "Higher success rates" link within the alert. If they click on the link in the banner alert they will get sent to: [What is 'Smart Transactions'?](https://support.metamask.io/transactions-and-gas/transactions/smart-transactions/) for more information. When returning to the confirmation they just navigated from the banner alert should not show and should never show again.

Edge Cases: 

If a user is new and setting up a wallet for the first time, they will not receive the Banner Alert. If a user imports a new wallet during a fresh install of the extension on a new browser or recovers a wallet, it's possible they may not see the alert if STX was on in a previous install. The STX Banner Alert is dismissed and will not show again if a user is in the state to get shown the banner and toggles STX off independently even if they do not physically dismiss the STX Banner Alert.

Migration Logic:

1. If `smartTransactionsOptInStatus` is `null` (new/never interacted)
 - Sets status to true
 - Enables notification flag
2. If status is false (previously opted out):
 - With no Ethereum Mainnet STX activity: Sets to true with notification
 - With existing Mainnet STX activity: Preserves user preference
3. If status is true: No changes needed

UI Components:

Implements `SmartTransactionsMigrationBanner` component for user notification.

The notification system bridges the migration changes with the UI, ensuring users are informed of the STX enablement while maintaining their ability to opt out through settings.

Target release: release-7.39.0
Affected user base: ~5.7M users who previously opted out of STX but have no STX activity.

## **Running Unit Tests**

Migration 067 test: 

```bash
yarn jest "./app/store/migrations/068.test.ts" --no-cache
```

SmartTransactions Migration Banner component test: 

```bash
yarn jest "./app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.test.ts" --no-cache
```

QuotesView component test: 

```bash
yarn jest "./app/components/UI/Swaps/QuotesView.test.ts" --no-cache
```

SendFlow component test: 

```bash
yarn jest "./app/components/Views/confirmations/SendFlow/Confirm/index.test.tsx" --no-cache
```

TransactionReview component test: 

```bash
yarn jest "./app/components/Views/confirmations/components/TransactionReview/index.test.tsx" --no-cach
```

## **Manual testing steps**

**Test Migration (using a wallet/account with no STX Transactions)**
Start with an older repo version:

**Terminal #1**

```bash
git checkout tags/v7.32.0
```
```bash
yarn setup && yarn watch
```

**Terminal #2**

```bash
yarn start:ios
```

1. Import or setup a wallet without STX transactions, launch the wallet (do not enable STX if prompted), check that toggle is OFF in: `Settings > Advanced > Smart Transactions`

2. Switch to feature branch and run app:

**Terminal #1**

```bash
git checkout feat/enable-stx-migration
```
```bash
yarn setup && yarn watch
```

**Terminal #2**

```bash
yarn start:ios
```

3. Test that Alert only shows during confirmation screens for transactions and contract interactions, but not for signing
4. Create a Send transaction to your own wallet for `0.0001` ETH
5. Ensure that Smart Transactions Banner Alert IS showing
6. Start a Swaps transaction on Ethereum Mainnet
7 Ensure that Smart Transactions Banner Alert IS showing
8. Try several Signs (ETH Sign, Personal Sign, Sign Typed Data, etc..) and ensure the STX Banner Alert does not show on those confirmations screens.
9. Ensure that the Smart Transactions Banner Alert is only showing on STX supported chains:
  - Ethereum
  - Sepolia
10. Ensure that link in alert "Higher success rates" link (inspect) goes to: [What is 'Smart Transactions'?](https://support.metamask.io/transactions-and-gas/transactions/smart-transactions/)
11. Dismiss the Smart Transactions Banner Alert 
12. Create a Send transaction to your own wallet for `0.0001` ETH
13. Ensure that Smart Transactions Banner Alert IS NOT showing

* Note: The Smart Transactions Banner Alert should not show up on chains like Linea which are not supported

## **Screenshots/Recordings**

The following before and after screenshots show the state of varying Mobile views like Swaps, Send, Contract Deployment, Signing, etc.. that show the `SmartTransactionsMigrationBanner` component and how it's displayed on each view. The before shows the view without the component being rendered and the after (after migration and before dismissal) show the component as it will appear for the user on each view.

### **Before**

<img width="150" alt="01-stx_before" src="https://github.com/user-attachments/assets/f526d21a-8a38-40ed-810e-1c102636ccbb" />

<img width="150" alt="02-send_before" src="https://github.com/user-attachments/assets/78155c14-295d-4d8a-94b6-6845158b29bf" />

<img width="150" alt="02-sendLegacy_before" src="https://github.com/user-attachments/assets/8f0049c5-ab00-45d0-adae-ed7d203ec79d" />

<img width="150" alt="04-signTypedDataV4_before" src="https://github.com/user-attachments/assets/07eef175-a431-4c88-a752-db8ea16c3328" />

<img width="150" alt="06-contractInteraction_before" src="https://github.com/user-attachments/assets/dcb63432-9f2d-4589-b6b6-25d5e13ea28f" />

<img width="150" alt="05-contractDeployment_before" src="https://github.com/user-attachments/assets/61dbc3f6-599b-435b-a994-07b1e1f41bcf" />

### **After**
<img width="150" alt="01-stx_after" src="https://github.com/user-attachments/assets/dae49867-093d-475f-a16c-911e79406e3f" />

<img width="150" alt="02-send_after" src="https://github.com/user-attachments/assets/5152d738-304d-4fa8-a835-84c4ba22878b" />

<img width="150" alt="02-sendLegacy_after" src="https://github.com/user-attachments/assets/84eebe4f-f374-48de-9f4a-76ab33da64f2" />

<img width="150" alt="04-signTypedDataV4_after" src="https://github.com/user-attachments/assets/c99b0251-2a9b-4f3a-83c3-a3ab5d6523ac" />

<img width="150" alt="06-contractInteraction_after" src="https://github.com/user-attachments/assets/09731034-a4cc-40f1-b5ff-23b2eccee0d5" />

<img width="150" alt="05-contractDeployment_after" src="https://github.com/user-attachments/assets/3c0ab0ba-2188-45e3-b0ae-622800f76773" />

<img width="150" alt="09-swap_after" src="https://github.com/user-attachments/assets/0943676d-eb21-4aa1-a975-3545d029bf34" />

<img width="150" alt="09-swap_after" src="https://github.com/user-attachments/assets/1b50c741-3105-4fb0-b010-ba08111ead89" />

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

httpJunkie committed:
wip: first stab at migration for stx by default.

httpJunkie committed:
feat: enable STX by default with migration 65:

- Enable STX for new users or those that have not enabled in settings
- Preserves settings for users who explicitly opted out and have existing STX transactions
- Adds a migration flag to track this change

httpJunkie committed:
fix: export State for test

httpJunkie committed:
fix: add explicit null checks (alongside `typeof`) in migration 65 error messages using team's established pattern.

httpJunkie committed:
test: create initial test for STX by default migration.

- Mock Sentry error handling ensuring we can track when errors are captured
- Setup `beforeEach()` restore/reset for all mocks
-  Define test cases for error handling
- Each case has: invalid input state, expected error message, and description
- Invalid States Test: how migration handles malformed data
- Runs each invalid state test case
- Verifies: Invalid state returns unchanged, Error is captured by Sentry, Error message matches expected format

Also, Setup first test to ensure STX is enabled and migration flag is set when undefined opt-in status

httpJunkie committed:
test: continued work on STX by default migration test.

- Null status test, similar to undefined opt-in status test
- Tests migration with null status
- Verifies that STX is enabled and migration flag is set

httpJunkie committed:
test: continued work on STX by default migration test

- create opt-out test
- tests that users explicit opt-out is respected when STX transactions exist
- verifies migration flag is set but preference is unchanged

httpJunkie committed:
test: continued work on STX by default migration test

- create opt-out test
- tests that users explicit opt-out is respected when STX transactions exist
- verifies migration flag is set but preference is unchanged

httpJunkie committed:
test: continued work on STX by default migration test

- tests that opt-out is overridden when user has no transaction history
- verifies migration flag is set and preference is changed

httpJunkie committed:
test: continued work on STX by default migration test

- added test for initializing preferences
- added test for when STX is already enabled
- use `merge` with `initialRootState` consistently across all tests

httpJunkie committed:
feat: create banner component for STX enabled

httpJunkie committed:
fix: comment out code that is not yet being used

httpJunkie committed:
feat: Updated string references to use new locale keys

httpJunkie committed:
feat: create useSmartTransactionsEnabled hook

- uses redux selectors to access the state
- checks flags we need (opt-in status and migration applied)
- Returns (isEnabled, isMigrationApplied) for our banner to use
- Provides proper TypeScript types

httpJunkie committed:
feat: implement test for hook

- check default states
- ensure proper state reading
- handle partial states

httpJunkie committed:
test: add test for SmartTransactionsEnabledBanner

- test conditional rendering
- check text content
- ensure close button functionality works
- ensure link functionality works

httpJunkie committed:
feat: update SmartTransactionsEnabledBanner and useTransactionsEnabled

- useSmartTransactionsEnabled enhanced to include dismiss functionality
- add a dispatch action to update the smartTransactionsOptInStatus
- remove external onClose prop from SmartTransactionsEnabledBanner in favor of using the hook's dismiss functionality
- update component and hook tests to account for new functionality

These changes require more changes in the next commit to add the smartTransactionsBannerDismissed field to our PreferencesController state as well to add a reducer case for SET_SMART_TRANSACTIONS_BANNER_DISMISSED and update any types/interfaces in our codebase that reference PreferencesController.

httpJunkie committed:
feat: add action to store settings

httpJunkie committed:
feat: add smartTransactionsBannerDismissed preferenceController and initialize it. Update the migration and test to account for the new change

httpJunkie committed:
feat: add SmartTransactionsEnabledBanner to TransactionReview

dan437 commented on app/actions/settings/index.js:
```diff
@@ -74,3 +74,10 @@ export function setTokenSortConfig(tokenSortConfig) {
   };
 }
+
+export function setSmartTransactionsBannerDismissed(bannerDismissed) {
```
nit: to make it clearer, I would call it `setSmartTransactionsMigrationBannerDismissed`

dan437 commented on app/actions/settings/index.js:
```diff
@@ -74,3 +74,10 @@ export function setTokenSortConfig(tokenSortConfig) {
+
+export function setSmartTransactionsBannerDismissed(bannerDismissed) {
+  return {
+    type: 'SET_SMART_TRANSACTIONS_BANNER_DISMISSED',
```
SET_SMART_TRANSACTIONS_MIGRATION_BANNER_DISMISSED

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsEnabledBanner/SmartTransactionsEnabledBanner.tsx:
```diff
@@ -0,0 +1,48 @@
+import { SmartTransactionsEnabledBannerProps } from './SmartTransactionsEnabledBanner.types';
+import useSmartTransactionsEnabled from '../../../../hooks/useSmartTransactionsEnabled/useSmartTransactionsEnabled';
+
+const SMART_TRANSACTIONS_LEARN_MORE = AppConstants.URLS.SMART_TXS;
```
Since this constant is only used in 1 place, you can just delete it and use it there directly. I would also rename it a little:
```
Linking.openURL(AppConstants.URLS.SMART_TRANSACTIONS);
```

dan437 commented on app/components/hooks/useSmartTransactionsEnabled/useSmartTransactionsEnabled.ts:
```diff
@@ -0,0 +1,61 @@
+
+const SET_SMART_TRANSACTIONS_BANNER_DISMISSED = 'SET_SMART_TRANSACTIONS_BANNER_DISMISSED';
+
+const useSmartTransactionsEnabled = () => {
```
`useSmartTransactionsEnabled` naming is very similar to existing [selectSmartTransactionsEnabled](https://github.com/MetaMask/metamask-mobile/blob/main/app/selectors/smartTransactionsController.ts#L14C14-L14C44), which can lead to confusion. 

If you see the [selectShouldUseSmartTransaction](https://github.com/MetaMask/metamask-mobile/blob/main/app/selectors/smartTransactionsController.ts#L46C14-L46C45) selector, it checks 2 things:
- isSmartTransactionsEnabled, which indicates if the STX feature is even available, no matter what's in Settings
- smartTransactionsOptInStatus, which indicates if it's on or off in Settings
Both have to be true in order for the mobile app to use smart transactions.

I see that in the `hooks` folder not many files use the `state: RootState` code. Either we should rename this hook since there is another function with a very similar name and different meaning or we can create individual selectors for isMigrationApplied, isBannerDismissed, etc. and remove this hook

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsEnabledBanner/SmartTransactionsEnabledBanner.styles.ts:
```diff
@@ -0,0 +1,12 @@
+import { StyleSheet, ViewStyle } from 'react-native';
```
Why don't we call this and other files `SmartTransactionsMigrationBanner`?

httpJunkie committed:
feat: increase migration number temporarily to avoid conflicts

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
feat: remove temporary migration file used to try and setup state, this idea did not pan out.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
feat: remove SmartTransactionsEnabledBanner from TransactionReview, ensure that styles are applied to match the mockup design and add the Banner Alter to the SendFlow Confirmation page.

httpJunkie committed:
feat: update SmartTransactionsEnabledBanner to dismiss ALert when link is viewed. Update useSmartTransactionsEnabled hook to set feature flags for smartTransactionsBannerDismissed. Remove action from app/actions/settings. Test is still failing (TBD)

httpJunkie committed:
fix: update useSmartTransactionsEnabled.test.ts

- update the Engine mock to be more explicit
- mock both the default export and the named exports of the Engine module

httpJunkie committed:
fix: update useSmartTransactionsEnabled hook to add additional checks to ensure:

- mock is clean at start
- mock was called exactly once
- arguments are correct

httpJunkie committed:
feat: update and improve useSmartTransactionsEnabled

- Make types in PreferencesState more accurate knowing the values will exist after migration
- Prevent unnecessary null checks and TS hints and error checking
- Memoize shouldShowBanner calculation to prevent unnecessary recalculations when other state changes, follows basic practices for derived state only recalculating when dependencies changes

httpJunkie committed:
fix: move smartTransactionsBannerDismissed to featureFlags

- Move smartTransactionsBannerDismissed from PreferencesController root to featureFlags object
- Add type safety improvements for feature flags initialization
- Add SmartTransactionsController state type guard
- Update migration test to check for banner state in featureFlags instead of root

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
feat: add banner alert to ApproveTransactionReview and TransactionReview

- covers token transfer approval
- covers dapp and contract interaction like deploy a contract and mint

httpJunkie committed:
feat: add SmartTransactionsEnabledBanner to Swaps Quotes VIew

httpJunkie committed:
fix: temporary code to reset `smartTransactionsBannerDismissed` for quicker manual testing of the conditional rendering.

httpJunkie committed:
feat: update QuickView and placement of the the SmartTransactionsEnableBanner to be within the `topBar` and `alertBar` view styles.

httpJunkie committed:
fix: update naming convention of SmartTransactionsEnableBanner on TransactionReview component reduce margin below banner and update all banners to have similar bottom margin

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: remove devTestId from TransactionReview view

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
fix: remove dev testing code from `useSmartTransactionsEnabled` hook.

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-mobile into feat/enable-stx-migration rename SmartTransactionsMigrationBanner

httpJunkie committed:
fix: refactor `useSmartTransactionsEnabled` hook to use selectors instead.

- Create new selectors file
- Update the SmartTransactionsMigrationBanner component to use these selectors directly
- Delete the hook file since its functionality would now be handled by selectors

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  const handleLearnMore = () => {
+    Linking.openURL(SMART_TRANSACTIONS_LEARN_MORE);
```
I would just use a constant directly:
```
Linking.openURL(AppConstants.URLS.SMART_TRANSACTIONS);
```

dan437 commented on app/components/Views/confirmations/components/TransactionReview/index.js:
```diff
@@ -496,6 +500,7 @@ class TransactionReview extends PureComponent {
   }
 
   renderTransactionReview = () => {
+    Logger.log('TransactionReview RENDER TEST');
```
We should remove this test code before merging.

httpJunkie replied:
I had missed that, thank you

httpJunkie committed:
refactor: move smartTransactionsMigrationApplied into PreferencesController featureFlags

- Move smartTransactionsMigrationApplied from PreferencesController root to featureFlags object
- Aligns with existing pattern of using featureFlags for UI states
- Maintains consistency with smartTransactionsBannerDismissed placement

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
fix: addressing Daniels comments

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,72 @@
+      }
+      onClose={dismissBanner}
+      style={styles.banner}
+      testID="smart-transactions-enabled-banner"
```
nit: I would call it after the component's name, in this case: `smart-transactions-migration-banner`

httpJunkie committed:
test: update snapshots for:

- TransactionReview
- Swaps Quote View

httpJunkie committed:
test: fix migration test

httpJunkie committed:
fix: update testID values after component name change to be consistent

httpJunkie committed:
test: update location of `smartTransactionsMigrationApplied` to be within featureFlags on component test

httpJunkie committed:
test: update snapshot for Approval to account for our component and styles added

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
feat: update migration to #067 from #096

httpJunkie committed:
fix: bump error message details to show 67 instead of 65 for the migration number

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
test: update test for migration to expect test #067 instead of #065

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of github.com:MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
feat: add my migration back after accepting merge change?

httpJunkie committed:
feat: Update SmartTransactionMigrationBanner chain support

use existing Mobile selectors for proper chain support validation. Ensure banner displays on supported chains only.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration plus "migration index comment"

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: simplify STX migration to ignore transaction history

Updates migration 67 to only check `smartTransactionOptInStatus` value when
determining if banner should be shown. Remove STX history check. Fixes test initialization states and expectations
to match new behavior.

httpJunkie committed:
test: update STX banner tests to use proper mocks

Updates SmartTransactionMigrationBanner tests to properly mock
selectors and fix TypeScript errors.

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: mock smart transaction selectors for tests

Tests are failing because `swapsState.featureFlags` is undefined. This is happening because our `SmartTransactionsMigrationBanner` component is being rendered inside `TransactionReview` tests, but the mock state doesn't include the necessary swaps state

httpJunkie committed:
fix: remove extra line in migrations index

httpJunkie committed:
test: fix test mocks for STX migration changes

Adds missing selector mocks for `SmartTransactionMigrationBanner` for ApproveView/Approve

httpJunkie committed:
test: add missing transaction simulation selector mock

Updates preferences selector mocks to include `useTransactionSimulations`

httpJunkie committed:
test: add missing selectors to component tests

Add selectUseTransactionSimulations mock to fix failing tests in Confirm,
TransactionReview and Approval components.

httpJunkie committed:
fix: update path in test and add `selectGasFeeEstimates`

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

metamaskbot committed:
Bump version number to 1545

httpJunkie committed:
fix: bump version to create testflight build

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

metamaskbot committed:
Bump version number to 1546

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: change package version back

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: remove change to project file

httpJunkie committed:
fix: revert changes to bitrise.yml

httpJunkie committed:
fix: revert build.gradle

httpJunkie committed:
Update migration to flip logic

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: update migration test after changes

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,90 @@
+    isEnabled &&
+    isMigrationApplied &&
+    !isBannerDismissed &&
+    isSmartTransactionsEnabled &&
```
`selectShouldUseSmartTransaction` already calls `selectSmartTransactionsEnabled`, so you don't need to check for it again here.

dan437 commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,90 @@
+    isMigrationApplied &&
+    !isBannerDismissed &&
+    isSmartTransactionsEnabled &&
+    shouldUseSmartTransaction,
```
`shouldUseSmartTransaction` requires API calls for feature flags and STX liveness to be made. This component is used in multiple places and I'm wondering if those API calls are always made when this component is used?

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
refactor(STX): clean up SmartTransactionsMigrationBanner by removing redundant STX state checks and simplify test setup

Removed from component:
- `selectSmartTransactionsEnabled`
- `selectSmartTransactionsOptInStatus`
- `isEnabled`

Removed from component test:
- `selectSmartTransactionsOptInStatus`
- `selectSmartTransactionsEnabled`
- `createMockState`

Add conditional container rendering to fix swaps quotes layout issue with fetch quotes message.

Match inherited styles in component test
Update QuotesView Snapshot

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

OGPoyraz commented on app/components/Views/confirmations/components/TransactionReview/__snapshots__/index.test.tsx.snap:
```diff
@@ -75,6 +75,14 @@ exports[`TransactionReview should match snapshot 1`] = `
                       }
                     }
                   >
+                    <View
+                      style={
+                        {
+                          "marginBottom": -8,
+                          "marginHorizontal": 16,
+                        }
+                      }
+                    />
```
Again we should avoid adding this `View` if there is nothing to show inside.

OGPoyraz commented on app/components/UI/Swaps/QuotesView.js:
```diff
@@ -148,6 +149,11 @@ const createStyles = (colors) =>
     },
+    smartTransactionsMigrationBanner: {
+      paddingHorizontal: 20,
+      marginBottom: -8,
```
Can we avoid adding negative margin values?

httpJunkie replied:
I have removed all use of negative margin. Even though similar components like Blockaid Banner have done so. There are no more negative margin use in this PR.

OGPoyraz replied:
Amazing job @httpJunkie thanks a lot

OGPoyraz commented on app/components/UI/Swaps/QuotesView.js:
```diff
@@ -1788,7 +1794,13 @@ function SwapsQuotesView({
+          </View>
+        )}
         {(!hasEnoughTokenBalance || !hasEnoughEthBalance) && (
+          <>
```
Is this fragment needed?

OGPoyraz commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.test.tsx:
```diff
@@ -0,0 +1,103 @@
+    mockedSmartTransactions.selectShouldUseSmartTransaction.mockReturnValue(true);
+  });
+
+  it('renders nothing when banner should be hidden', () => {
```
```suggestion
it('renders nothing when banner is dismissed', () => {
```

httpJunkie replied:
I will update this test and add a few extra. I think you make a good point and now that I'm thinking of it. I want to cover more. I will commit those shortly.

httpJunkie replied:
updated

OGPoyraz commented on app/components/Views/confirmations/components/TransactionReview/index.test.tsx:
```diff
@@ -48,6 +48,21 @@ jest.mock(
+  selectShouldUseSmartTransaction: () => false,
+}));
+
+jest.mock('../../../../../reducers/swaps', () => ({
+  swapsStateSelector: () => ({
+    featureFlags: {
+      smart_transactions: {
+        mobile_active: false
+      }
+    }
+  })
+}));
```
If I am not mistaken this mocks are unnecessary additions in this test file.

httpJunkie replied:
As long as I am adding the banner to this component, I will need these mocks. Without them we get a series of failed tests:

```
 FAIL  app/components/Views/confirmations/components/TransactionReview/index.test.tsx (56.579 s)
  TransactionReview
    ✓ should render correctly (5 ms)
    ✕ should match snapshot (318 ms)
    ✕ should display blockaid banner (144 ms)
    ✕ should have enabled confirm button if from account has balance (138 ms)
    ✕ should not have confirm button disabled if from account has no balance and also if there is no error (54 ms)
    ✕ should have confirm button disabled if error is defined (41 ms)
```

OGPoyraz replied:
My mistake, I thought I could run them without it, thanks for having a second look

OGPoyraz commented:
![Screenshot 2025-02-07 at 14 16 09](https://github.com/user-attachments/assets/0cc54d96-bf1a-4604-8616-e16c9ce0eaa3)

The banner looks a bit overflow in this container, is this the expected design?

httpJunkie commented:
> The banner looks a bit overflow in this container, is this the expected design?

I definitely understand the concern. This banner does not have any additional or override CSS to limit it's width within the containers it is in. I don't have a screenshot here but I can add one. Let's say that we were to interact with a Dapp trying to deploy a malicious contract. There might be another Banner alert that shows stacked on top of this one. I would say that the STX Banner Alert is the same size and rendered in the same place. These components that we are displaying in have only one place where these banner alerts show. I have made sure that tis banner displays in the same way as any other banner that would display on that view. I have noticed that the elements like From and To have additional margin around them. I don't think it's overflow, I think it is just taking up the full space within that container and other UI elements have different styling.

I can look further into it.

httpJunkie commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,82 @@
+            {strings('smart_transactions_migration.link')}
+          </Text>
+          <Text style={styles.description}>
+            {' '}{strings('smart_transactions_migration.description')}
```
yes

httpJunkie committed:
test: expand banner visibility test coverage

- Rename original test to be more specific
- Add tests for STX enabled and migration status
- Group tests into visibility, interactions, and styling
- Ensure comprehensive coverage of banner display conditions

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie commented on app/components/Views/confirmations/SendFlow/Confirm/styles.ts:
```diff
@@ -153,6 +153,10 @@ const createStyles = (colors: any) =>
       marginBottom: 24,
     },
+    smartTransactionsMigrationBanner: {
+      marginBottom: -8,
```
It was added here to ensure proper spacing only needed for this page. When we remove this feature (Banner Alert) It's style will be removed also so it seems to be better to add it here rather than adjust the CSS from within the component only when it's applied to this page.

httpJunkie committed:
fix: enable custom style override and update test

- Uncomment style spread in banner styles
- Update test to expect custom style to override defaults

httpJunkie committed:
fix: apply marginVertical and ensure test handles properly

httpJunkie commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.test.tsx:
```diff
@@ -0,0 +1,103 @@
+
+jest.mock('../../../../../selectors/smartTransactionsController', () => ({
+  selectShouldUseSmartTransaction: jest.fn(() => true),
+}));
```
I just saw this message, I could complete this request in a few hours as I'm at a stopping pint right now, but the reason I did it this way was because each selector has deep dependencies, making direct state construction complex. 

I can complete the state mocking approach by adding all required state or leave as-is.

OGPoyraz replied:
I definitely understand that these selectors have deep dependencies and this is exactly why we need to handle that in these test rather than mocking them. 

You could create some states something like that we use in confirmations (of course for STX)
https://github.com/MetaMask/metamask-mobile/blob/main/app/util/test/confirm-data-helpers.ts

Then you can re-use them later 
https://github.com/MetaMask/metamask-mobile/blob/main/app/components/Views/confirmations/components/Confirm/AccountNetworkInfo/AccountNetworkInfo.test.tsx#L13

I know it looks like a painful at first look, but doing that will give us more confidence of the code that we import confirmations. We had bunch of issues appearing from out of nowhere just because these implementations heavily in the past.

OGPoyraz replied:
We agreed on creating follow up PR for this.

httpJunkie committed:
fix: move `f (!shouldShowBanner)` above dismissBanner const

httpJunkie commented on app/components/Views/confirmations/Approval/__snapshots__/index.test.tsx.snap:
```diff
@@ -592,6 +592,14 @@ exports[`Approval render matches snapshot 1`] = `
                                                     }
                                                   }
                                                 >
+                                                  <View
+                                                    style={
+                                                      {
+                                                        "marginBottom": -8,
+                                                        "marginHorizontal": 16,
+                                                      }
+                                                    }
+                                                  />
```
The PR will not pass without it, this is a snapshot update?

OGPoyraz replied:
Apologies for the confusion @httpJunkie , if this is added in the snapshots, this mean that there is a posibility to include this empty `View` with negative margin for some cases. 

This is an unwanted empty component, because this could broke layout because it's empty and adding negative margin.

OGPoyraz replied:
Closing this as it's not putting an empty <View/> anymore.

httpJunkie committed:
fix: remove unused fragment

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

OGPoyraz commented:
> I definitely understand the concern. This banner does not have any additional or override CSS to limit it's width within the containers it is in. I don't have a screenshot here but I can add one. Let's say that we were to interact with a Dapp trying to deploy a malicious contract. There might be another Banner alert that shows stacked on top of this one. I would say that the STX Banner Alert is the same size and rendered in the same place. These components that we are displaying in have only one place where these banner alerts show. I have made sure that tis banner displays in the same way as any other banner that would display on that view. I have noticed that the elements like From and To have additional margin around them. I don't think it's overflow, I think it is just taking up the full space within that container and other UI elements have different styling.
> 
> I can look further into it.

@httpJunkie it would be awesome if we could address this overflow, I am 100% will be return us back as a visual bug report

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie commented:
> it would be awesome if we could address this overflow, I am 100% will be return us back as a visual bug report



> > I definitely understand the concern. This banner does not have any additional or override CSS to limit it's width within the containers it is in. I don't have a screenshot here but I can add one. Let's say that we were to interact with a Dapp trying to deploy a malicious contract. There might be another Banner alert that shows stacked on top of this one. I would say that the STX Banner Alert is the same size and rendered in the same place. These components that we are displaying in have only one place where these banner alerts show. I have made sure that tis banner displays in the same way as any other banner that would display on that view. I have noticed that the elements like From and To have additional margin around them. I don't think it's overflow, I think it is just taking up the full space within that container and other UI elements have different styling.
> > I can look further into it.
> 
> @httpJunkie it would be awesome if we could address this overflow, I am 100% will be return us back as a visual bug report

I can address in this PR, however I will note that there are at least two other banners I have tested and they both overflow, here is one of them. You can test this by going to our test dapp and connecting and selecting "Malicious Transfer" or "Malicious Trade". When creating this component I actually relied on this banner to show me that my banner was showing up like the others.

<video src="[LINK](https://github.com/user-attachments/assets/865f0070-9ae1-4c12-bfd2-bbe26c092134)" controls="controls" style="max-width: 730px;">
</video>

Although I will see that mine is not overriding and overall looks correct within the ui. This is my first contribution to mobile and I thought that using the existing banner and following the rendering of previous banners would be ok. LMKWYT. I'm open to fixing it. But it feels like we will have inconsistent banner displays if I change mine. Because my banner sometimes shows up with the others... Stacked. I will provide an example of that. If I can fix them all with a fix. That might be a huge win?

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: Moved `dismissBanner` useCallback before the conditional return

httpJunkie committed:
fix: remove negative margin on Banner Alert view styles

httpJunkie committed:
fix: update snapshots that expected prior negative margin

httpJunkie commented on app/components/Views/confirmations/components/SmartTransactionsMigrationBanner/SmartTransactionsMigrationBanner.tsx:
```diff
@@ -0,0 +1,82 @@
+    }
+  }, []);
+
+  if (!shouldShowBanner) {
+    return null;
+  }
```
fixed

httpJunkie committed:
fix: conditionally render view and component so that the view does not display without the rendering of the component and update snapshot

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: change to camelCase

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie commented on app/components/Views/confirmations/components/TransactionReview/index.js:
```diff
@@ -117,6 +117,9 @@ const createStyles = (colors) =>
       marginHorizontal: 16,
       marginBottom: -8,
     },
+    SmartTransactionsMigrationBanner: {
```
fixed, ty

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
fix: conflict resolution

httpJunkie committed:
Merge branch 'feat/enable-stx-migration' of https://github.com/MetaMask/metamask-mobile into feat/enable-stx-migration

httpJunkie committed:
fix: migration conflict rosolution

httpJunkie committed:
fix: merge conflict remove space

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration

httpJunkie committed:
Merge branch 'main' into feat/enable-stx-migration
